README.md
Audit-DBI-TT2
=============

[![Build Status](https://travis-ci.org/guillaumeaubert/Audit-DBI-TT2.png?branch=master)](https://travis-ci.org/guillaumeaubert/Audit-DBI-TT2)
[![Coverage Status](https://coveralls.io/repos/guillaumeaubert/Audit-DBI-TT2/badge.png?branch=master)](https://coveralls.io/r/guillaumeaubert/Audit-DBI-TT2?branch=master)

A Template Toolkit plugin to display audit events recorded by [Audit::DBI]
(https://metacpan.org/release/Audit-DBI).


INSTALLATION
------------

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install


SUPPORT AND DOCUMENTATION
-------------------------

After installing, you can find documentation for this module with the
perldoc command.

	perldoc Audit::DBI::TT2


You can also look for information at:

 * [GitHub's request tracker (report bugs here)]
   (https://github.com/guillaumeaubert/Audit-DBI-TT2/issues)

 * [AnnoCPAN, Annotated CPAN documentation]
   (http://annocpan.org/dist/Audit-DBI-TT2)

 * [CPAN Ratings]
   (http://cpanratings.perl.org/d/Audit-DBI-TT2)

 * [MetaCPAN]
   (https://metacpan.org/release/Audit-DBI-TT2)


LICENSE AND COPYRIGHT
---------------------

Copyright (C) 2010-2014 Guillaume Aubert

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/



Changes
Revision history for Audit-DBI-TT2

v2.2.5  2014-04-20
        - Fixed "package version matches dist version" for CPANTS.
        - Fixed license declared in meta files.
        - General cleanup.

v2.2.4  2013-06-15
        - Fixed test conditional.

v2.2.3  2013-06-14
        - Fall back to a simpler test on some architectures with a non-working
          tzset().
        - Added Coveralls.io support and badge.

v2.2.2  2013-06-02
        - Fixed conditional.

v2.2.1  2013-06-01
        - Improved test to detect if tzset() is available.

v2.2.0  2013-05-28
        - Replaced Data::Dumper with Data::Dump.
        - Skipped time formatting test for platforms without POSIX::tzset().

v2.1.1  2013-05-21
        - Use fixed indentation in tests, to be able to compare the output.

v2.1.0  2013-05-17
        - Removed Class::Date dependency, replaced by POSIX functions.

v2.0.0  2013-05-14
        - Split Audit::DBI::TT2 off the Audit-DBI distribution, as it adds
          several large dependencies but it is not necessary for the use of
          the audit system itself.
        - Added tests for all the subroutines.

Important: this distribution was previously bundled as part of Audit::DBI, and
was split in v1.8.0. To prevent issues with reindexing in CPAN, Audit::DBI::TT2
starts at v2.0.0.


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.006;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Audit::DBI::TT2',
  'VERSION_FROM' => 'lib/Audit/DBI/TT2.pm',
  'PREREQ_PM' => {
                   'Audit::DBI' => '1.008001',
                   'DBD::SQLite' => 0,
                   'DBI' => 0,
                   'Data::Dump' => 0,
                   'HTML::Entities' => 0,
                   'POSIX' => 0,
                   'Scalar::Util' => 0,
                   'Template::Plugin' => 0,
                   'Template::Stash' => 0,
                   'Test::Exception' => 0,
                   'Test::FailWarnings' => 0,
                   'Test::More' => '0.94'
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {},
  'test' => {
              'TESTS' => 't/*.t'
            }
)
;


lib/Audit/DBI/TT2.pm
package Audit::DBI::TT2;

use strict;
use warnings;

use base 'Template::Plugin';

use Data::Dump qw();
use HTML::Entities qw();
use POSIX qw();
use Template::Stash;


=head1 NAME

Audit::DBI::TT2 - A Template Toolkit plugin to display audit events recorded by L<Audit::DBI>.


=head1 VERSION

Version 2.2.5

=cut

our $VERSION = '2.2.5';


=head1 SYNOPSIS

In your Perl code:

	use Audit::DBI::TT2;
	use Template;

	my $template = Template->new(
		{
			PLUGINS =>
			{
				audit => 'Audit::DBI::TT2',
			},
		}
	) || die $Template::ERROR;

In your TT2 template:

	[% USE audit %]
	[% FOREACH result IN audit.format_results( results ) %]
		...
	[% END %]

Note: a fully operational example of a search interface for Audit::DBI events
using this module for the display of the results is available in the
C<examples/> directory of this distribution.


=head1 FUNCTIONS

=head2 format_results()

Format the following fields for display as HTML:

=over 4

=item * diff

(accessible as diff_formatted)

=item * information

(accessible as information_formatted)

=item * event_time

(accessible as event_time_formatted)

=back

	[% FOREACH result IN audit.format_results( results ) %]
		<div>
			Formatted information: [% result.information_formatted %]<br/>
			Formatted diff: [% result.diff_formatted %]<br/>
			Formatted event time: [% result.event_time_formatted %]
		</div>
	[% END %]

=cut

sub format_results
{
	my ( $self, $results ) = @_;

	foreach my $result ( @$results )
	{
		$result->{information_formatted} = html_dumper( $result->get_information() );
		$result->{diff_formatted} = html_dumper( $result->get_diff() );
		$result->{event_time_formatted} = POSIX::strftime(
			"%Y-%m-%d %H:%M:%S",
			POSIX::localtime( $result->{event_time} ),
		);
	}

	return $results;
}


=head2 html_dumper()

Format a data structure for display as HTML.

	my $formatted_data = Audit::DBI::TT2::html_dumper( $data );

=cut

sub html_dumper
{
	my ( $data ) = @_;
	return undef
		if !defined( $data );

	my $string = Data::Dump::dump( $data );
	$string = HTML::Entities::encode_entities( $string );
	$string =~ s/ /&nbsp;/g;
	$string =~ s/\n/<br\/>/g;

	return $string;
}


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/Audit-DBI-TT2/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Audit::DBI::TT2


You can also look for information at:

=over 4

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/Audit-DBI-TT2/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Audit-DBI-TT2>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Audit-DBI-TT2>

=item * MetaCPAN

L<https://metacpan.org/release/Audit-DBI-TT2>

=back


=head1 AUTHOR

L<Guillaume Aubert|https://metacpan.org/author/AUBERTG>, C<< <aubertg at cpan.org> >>.


=head1 ACKNOWLEDGEMENTS

I originally developed this project for ThinkGeek
(L<http://www.thinkgeek.com/>). Thanks for allowing me to open-source it!


=head1 COPYRIGHT & LICENSE

Copyright 2010-2014 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;


examples/js/search.js
/**
 * Audit::DBI::TT2 v2.2.5
 * https://metacpan.org/release/Audit-DBI
 *
 * Copyright 2010-2014 Guillaume Aubert
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see http://www.gnu.org/licenses/
 */


/**
 * Add a new row in the list of criteria.
 *
 * @param {String} invert - null to search on the criteria, not null to search on its opposite.
 * @param {String} criteria - the type of criteria.
 * @param {String} values - the values associated with this criteria.
 */
var count = 0;
function add_new_criteria(invert, criteria, values)
{
	// Copy the row and set a unique ID on it.
	$('#criteria').append(
		$('#model_row').html()
	);
	$('#criteria .row:last').attr('id', 'row_' + count);

	// Set unique names to the various inputs.
	$('#criteria .row:last .remove input').bind(
		'click',
		function(e) {
			$('#' + $(this).parent().parent().attr('id') ).remove();
		}
	);

	// If criteria and invert are passed, set them.
	if (invert != null) {
		$('#criteria .row:last .invert select').val(invert);
	}
	if (criteria != null) {
		$('#criteria .row:last .criteria select').val(criteria);
	}
	if (values != null) {
		show_values_div($('#criteria .row:last'), criteria, values);
	}

	// Set up the DatePicker fields for that row.
	$("#row_" + count + " .datepicker").datepicker();

	// Add logic to display the different "values" DIVs based on the criteria
	// selected.
	$('#criteria .row:last .criteria select').bind(
		'change',
		function(e)
		{
			var criteria = $(this).val();
			var container = $(this).parent().parent();
			show_values_div(container, criteria);
		}
	);
	count++;
}


/**
 * Show the UI to select the values for a given criteria.
 *
 * @param {String} container - a jQuery object containing the HTML for the criteria.
 * @param {String} criteria - the type of criteria.
 * @param {String} values - the values associated with this criteria.
 */
function show_values_div(container, criteria, values)
{
	var value;
	if (values != null) {
		value = values.split('::');
	} else {
		value = new Array();
	}
	container.find('.values').css('display', 'none');

	// Display IP address search.
	if ( criteria == 'ip_address' ) {
		container.find('.values_ip_address').css('display', 'block');
		container.find('.values_ip_address input').val(value[0]);

	// Display date range search.
	} else if ( criteria == 'date_range' ) {
		container.find('.values_date_range').css('display', 'block');
		var from = container.find('.values_date_range input:first');
		from.val(value[0]);
		//from.datepicker();
		var to = container.find('.values_date_range input:last');
		to.val(value[1]);
		//to.datepicker();

	// Display subject type search.
	} else if ( criteria == 'subject_type' ) {
		container.find('.values_subject_type').css('display', 'block');
		container.find('.values_subject_type input:first').val(value[0]);
		container.find('.values_subject_type input:last').val(value[1]);

	// Display event search.
	} else if ( criteria == 'event' ) {
		container.find('.values_event').css('display', 'block');
		container.find('.values_event input').val(value[0]);

	// Display account logged in search.
	} else if ( criteria == 'account_logged_in' ) {
		container.find('.values_account_logged_in').css('display', 'block');
		container.find('.values_account_logged_in input').val(value[0]);

	// Display account affected search.
	} else if ( criteria == 'account_affected' ) {
		container.find('.values_account_affected').css('display', 'block');
		container.find('.values_account_affected input').val(value[0]);

	// Display indexed data search.
	} else if ( criteria == 'indexed_data' ) {
		container.find('.values_indexed_data').css('display', 'block');
		container.find('.values_indexed_data input:first').val(value[0]);
		container.find('.values_indexed_data input:last').val(value[1]);
	}
}


/**
 * Make a URL representing the parameters chosen in the UI.
 */
function submit_search()
{
	var url = '?action=results';
	jQuery.each(
		$('#criteria .row'),
		function()
		{
			// Find the various values for the search.
			var invert = $(this).find('.invert select').val();
			var criteria = $(this).find('.criteria select').val();
			if (criteria == '') return;
			url += '&' + escape(criteria + invert) + '=';

			// Find the specific search values.
			if ( criteria == 'ip_address' ) {
				var ip_addresses = $(this).find('.values_ip_address input').val();
				url += escape(ip_addresses);

			} else if ( criteria == 'date_range' ) {
				var from = $(this).find('.values_date_range input:first').val();
				var to = $(this).find('.values_date_range input:last').val();
				url += escape(from) + '::' + escape(to);

			} else if ( criteria == 'subject_type' ) {
				var type = $(this).find('.values_subject_type input:first').val();
				var IDs = $(this).find('.values_subject_type input:last').val();
				url += escape(type) + '::' + escape(IDs);

			} else if ( criteria == 'event' ) {
				var event = $(this).find('.values_event input').val();
				url += escape(event);

			} else if ( criteria == 'account_logged_in' ) {
				var account_logged_in = $(this).find('.values_account_logged_in input').val();
				url += escape(account_logged_in);

			} else if ( criteria == 'account_affected' ) {
				var account_affected = $(this).find('.values_account_affected input').val();
				url += escape(account_affected);

			} else if ( criteria == 'indexed_data' ) {
				var type = $(this).find('.values_indexed_data input:first').val();
				var values = $(this).find('.values_indexed_data input:last').val();
				url += escape(type) + '::' + escape(values);
			}
		}
	);

	window.location.href = url;
}


/*
 * After loading, populate criteria with input if available.
 */
$(document).ready(
	function()
	{
		// Check for parameters and populate the criteria accordingly.
		var rows_added = false;
		var params = window.location.search.substring(1).split('&');
		jQuery.each(
			params,
			function()
			{
				var temp = this.split('=');
				if (temp[0] == '' || temp[0] == 'action') return;

				var invert = temp[0].substr(-1);
				var criteria = temp[0].substr(0, temp[0].length-1);

				add_new_criteria(invert, criteria, unescape(temp[1]));
				rows_added = true;
			}
		);

		// If no row was created, create an empty row.
		if (!rows_added) {
			add_new_criteria();
		}
	}
);



examples/js/results.js
/**
 * Audit::DBI::TT2 v2.2.5
 * https://metacpan.org/release/Audit-DBI
 *
 * Copyright 2010-2014 Guillaume Aubert
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see http://www.gnu.org/licenses/
 */
$(document).ready(
	function()
	{
		// Search all the links in code columns.
		$('#results .code a').each(
			function()
			{
				// Activate toggling.
				$(this).click(
					function()
					{
						$(this).parent().find('div').toggle();
						return false;
					}
				);
			}
		);

		// Activate toggling all
		$('#toggle_all').click(
			function()
			{
				$('#results .code div').each(
					function()
					{
						$(this).toggle();
					}
				);
				return false;
			}
		);
	}
);



Build.PL
use 5.006;
use strict;
use warnings;

use Module::Build;


my $class = Module::Build->subclass(
	code =>
	q|
		# Override 'test' to support running extended tests in the xt/
		# directory.
		sub ACTION_test
		{
			my ( $self ) = @_;

			$self->recursive_test_files(1);

			$self->test_files( 't', 'xt' )
				if $ENV{'RELEASE_TESTING'};

			return $self->SUPER::ACTION_test();
		}

		# Force running extended tests when testing the distribution.
		sub ACTION_disttest
		{
			my ( $self ) = @_;

			local $ENV{ RELEASE_TESTING } = 1;

			return $self->SUPER::ACTION_disttest();
		}
	|,
);

my $builder = $class->new(
	module_name          => 'Audit::DBI::TT2',
	dist_abstract        => 'A Template Toolkit plugin to display audit events recorded by Audit::DBI',
	license              => 'GPL_3',
	dist_author          => q{Guillaume Aubert <aubertg@cpan.org>},
	dist_version_from    => 'lib/Audit/DBI/TT2.pm',
	build_requires       =>
	{
		'DBD::SQLite'          => 0,
		'DBI'                  => 0,
		'Scalar::Util'         => 0,
		'Test::Exception'      => 0,
		'Test::FailWarnings'   => 0,
		'Test::More'           => 0.94,
	},
	requires             =>
	{
		'Audit::DBI'           => 1.008001,
		'Data::Dump'           => 0,
		'HTML::Entities'       => 0,
		'POSIX'                => 0,
		'Template::Plugin'     => 0,
		'Template::Stash'      => 0,
		'perl'                 => 5.006,
	},
	add_to_cleanup       =>
	[
		'Audit-DBI-TT2-*',
	],
	create_makefile_pl   => 'traditional',
	configure_requires   =>
	{
		'Module::Build'        => 0,
	},
	meta_merge           =>
	{
		resources              =>
		{
			repository  => 'https://github.com/guillaumeaubert/Audit-DBI-TT2',
			homepage    => 'https://metacpan.org/release/Audit-DBI-TT2',
			bugtracker  => 'https://github.com/guillaumeaubert/Audit-DBI-TT2/issues',
		},
	},
	recursive_test_files => 1,
);

$builder->create_build_script();
